File: Kotlin/ex1.kt

fun main() {
    val input = readLine() ?: return
    val result = ex1(input)
    result.forEach { println(it) }
}

fun ex1(input: String): List<String> {
    var substr = ""
    val res = mutableListOf<String>()
    var state = 0

    for (i in input.indices) {
        when (input[i]) {
            '1' -> {
                when (state) {
                    0 -> {
                        state = 1
                        substr += '1'
                    }
                    1 -> {
                        state = 2
                        substr += '1'
                    }
                    3 -> {
                        substr += '1'
                        res.add(substr)
                        substr = ""
                        state = 0
                    }
                    else -> {
                        state = 0
                        substr = ""
                    }
                }
            }
            '0' -> {
                when (state) {
                    1, 2, 3 -> {
                        state = 3
                        substr += '0'
                    }
                    else -> {
                        state = 0
                        substr = ""
                    }
                }
            }
            else -> {
                state = 0
                substr = ""
            }
        }
    }

    return res
}

File: C#/ex1.cs
using System;
using System.Collections.Generic;
using System.Text;
class Program
{
    static List<string> Ex1(string input)
    {
        StringBuilder substr = new StringBuilder();
        List<string> res = new List<string>();
        int state = 0;

        foreach (char ch in input)
        {
            if (ch == '1')
            {
                if (state == 0)
                {
                    state = 1;
                    substr.Append('1');
                }
                else if (state == 1)
                {
                    state = 2;
                    substr.Append('1');
                }
                else if (state == 3)
                {
                    substr.Append('1');
                    res.Add(substr.ToString());
                    substr.Clear();
                    state = 0;
                }
                else
                {
                    state = 0;
                    substr.Clear();
                }
            }
            else if (ch == '0')
            {
                if (state == 1 || state == 2 || state == 3)
                {
                    state = 3;
                    substr.Append('0');
                }
                else
                {
                    state = 0;
                    substr.Clear();
                }
            }
            else
            {
                state = 0;
                substr.Clear();
            }
        }
        return res;
    }

    static void Main()
    {
        string input = Console.ReadLine();
        List<string> result = Ex1(input);

        foreach (string item in result)
        {
            Console.WriteLine(item);
        }
    }
}
